### DISCUSSION TITLE
How to build / embed sqlite extensions

### DISCUSSION THREAD

**Discussion by Sytten** [rusqlite/rusqlite#1125] [Questions] [answered]
Hi!
I am trying to build & bundle the SHA1 extension, but I am not sure where to start.
I use the bundled version of `libsqlite3-sys` and I also want to statically link the extension.
Is there a guide / directions someone could give me on that?
Thanks

**Comment by Sytten** [comment, marked as answer]
I managed to get working. I decided to create my own library on top of `libsqlite3-sys`.
You can take inspiration: https://github.com/caido/libsqlite3

The basic idea, you need 3 things:

**build.rs**
```rust
fn main() {
    println!("cargo:rerun-if-changed=sqlite3/sha1.c");
    cc::Build::new().file("sqlite3/sha1.c").compile("sha1");
}
```

**bindings.rs**
```rust
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct sqlite3([u8; 0]);

#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct sqlite3_api_routines([u8; 0]);

extern "C" {
    pub fn sqlite3_sha_init(
        db: *mut sqlite3,
        pzErrmsg: *mut *mut ::std::os::raw::c_char,
        pApi: *const sqlite3_api_routines,
    );
}
```

**lib.rs**
```rust
pub fn init() {
    unsafe {
        let ptr = bindings::sqlite3_sha_init as *const ();
        sqlite3_auto_extension(Some(
            std::mem::transmute::<*const (), unsafe extern "C" fn()>(ptr),
        ));
    }
}
```

Then you just need to call `init` once at the start of your program and you are good to go!

**Comment by gwenn** [comment]
https://sqlite.org/c3ref/auto_extension.html ?
